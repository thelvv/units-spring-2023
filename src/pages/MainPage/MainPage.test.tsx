import React from 'react';

import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MainPage } from './MainPage';
import { getNextSortBy } from '../../utils/getNextSortBy';
jest.mock('../../hooks/useCurrentTime');
jest.mock('../../utils/getPrice');
jest.mock('../../utils/getNextSortBy');

afterEach(jest.clearAllMocks);
describe('MainPage render', () => {
    it('should render correctly', () => {
        const rendered = render(<MainPage />);

        expect(rendered.asFragment()).toMatchSnapshot();
    });

    it('Should sort after button click', () => {
        const rendered = render(<MainPage />);
        const sortButton = rendered.getByTestId('sort-button');
        fireEvent.click(sortButton);
        expect(rendered).toMatchSnapshot();
    });

    it('Should call function getNextSortBy after sort button click', () => {
        const rendered = render(<MainPage />);
        const sortButton = rendered.getByTestId('sort-button');
        expect(getNextSortBy).toHaveBeenCalledTimes(0);
        fireEvent.click(sortButton);
        expect(getNextSortBy).toHaveBeenCalledTimes(1);
    });

    const categories = ['Электроника', 'Для дома', 'Одежда'];
    test.each(categories)('should filter by %category category', (category) => {
        const rendered = render(<MainPage />);
        const rendered_category = rendered
            .getAllByText(category)
            .filter((item) => item.classList.contains('categories__badge'))[0];
        fireEvent.click(rendered_category);

        expect(rendered).toMatchSnapshot();
    });

    test.each(categories)(
        'should get a new class after being clicked',
        (category) => {
            const rendered = render(<MainPage />);
            const rendered_category = rendered
                .getAllByText(category)
                .filter((item) => item.classList.contains('categories__badge'))[0];
            fireEvent.click(rendered_category);

            expect(rendered_category).toHaveClass('categories__badge_selected');
        }
    );
});
