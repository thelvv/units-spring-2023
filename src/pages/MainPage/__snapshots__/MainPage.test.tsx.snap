// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage render Should sort after button click 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="main-page"
      >
        <h2
          class="main-page__title"
        >
          VK Маркет
        </h2>
        <h3 />
        <div
          class="main-page__parameters"
        >
          <div
            class="categories"
          >
            <div
              class="categories__badge"
            >
              Одежда
            </div>
            <div
              class="categories__badge"
            >
              Для дома
            </div>
            <div
              class="categories__badge"
            >
              Электроника
            </div>
          </div>
          <button
            class="sort-button"
            data-testid="sort-button"
          >
            Сортировать 
          </button>
        </div>
        <div
          class="product-card"
        >
          <div
            class="product-card__text"
          >
            <h2
              class="product-card__name"
            >
              Настольная лампа
            </h2>
            <div
              class="product-card__description"
            >
              Говорят, что ее использовали в pixar
            </div>
            <h1
              class="product-card__price"
            />
            <div
              class="product-card__category"
            >
              Для дома
            </div>
          </div>
          <img
            alt="Настольная лампа"
            class="product-card__image"
            data-testid="product-card__image"
            src="/lamp.png"
          />
        </div>
        <div
          class="product-card"
        >
          <div
            class="product-card__text"
          >
            <h2
              class="product-card__name"
            >
              Костюм гуся
            </h2>
            <div
              class="product-card__description"
            >
              Запускаем гуся, работяги
            </div>
            <h1
              class="product-card__price"
            />
            <div
              class="product-card__category"
            >
              Одежда
            </div>
          </div>
        </div>
        <div
          class="product-card"
        >
          <div
            class="product-card__text"
          >
            <h2
              class="product-card__name"
            >
              Принтер
            </h2>
            <div
              class="product-card__description"
            >
              Незаменимая вещь для студента
            </div>
            <h1
              class="product-card__price"
            />
            <div
              class="product-card__category"
            >
              Электроника
            </div>
          </div>
        </div>
        <div
          class="product-card"
        >
          <div
            class="product-card__text"
          >
            <h2
              class="product-card__name"
            >
              IPhone 14 Pro
            </h2>
            <div
              class="product-card__description"
            >
              Latest iphone, buy it now
            </div>
            <h1
              class="product-card__price"
            />
            <div
              class="product-card__category"
            >
              Электроника
            </div>
          </div>
          <img
            alt="IPhone 14 Pro"
            class="product-card__image"
            data-testid="product-card__image"
            src="/iphone.png"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="main-page"
    >
      <h2
        class="main-page__title"
      >
        VK Маркет
      </h2>
      <h3 />
      <div
        class="main-page__parameters"
      >
        <div
          class="categories"
        >
          <div
            class="categories__badge"
          >
            Одежда
          </div>
          <div
            class="categories__badge"
          >
            Для дома
          </div>
          <div
            class="categories__badge"
          >
            Электроника
          </div>
        </div>
        <button
          class="sort-button"
          data-testid="sort-button"
        >
          Сортировать 
        </button>
      </div>
      <div
        class="product-card"
      >
        <div
          class="product-card__text"
        >
          <h2
            class="product-card__name"
          >
            Настольная лампа
          </h2>
          <div
            class="product-card__description"
          >
            Говорят, что ее использовали в pixar
          </div>
          <h1
            class="product-card__price"
          />
          <div
            class="product-card__category"
          >
            Для дома
          </div>
        </div>
        <img
          alt="Настольная лампа"
          class="product-card__image"
          data-testid="product-card__image"
          src="/lamp.png"
        />
      </div>
      <div
        class="product-card"
      >
        <div
          class="product-card__text"
        >
          <h2
            class="product-card__name"
          >
            Костюм гуся
          </h2>
          <div
            class="product-card__description"
          >
            Запускаем гуся, работяги
          </div>
          <h1
            class="product-card__price"
          />
          <div
            class="product-card__category"
          >
            Одежда
          </div>
        </div>
      </div>
      <div
        class="product-card"
      >
        <div
          class="product-card__text"
        >
          <h2
            class="product-card__name"
          >
            Принтер
          </h2>
          <div
            class="product-card__description"
          >
            Незаменимая вещь для студента
          </div>
          <h1
            class="product-card__price"
          />
          <div
            class="product-card__category"
          >
            Электроника
          </div>
        </div>
      </div>
      <div
        class="product-card"
      >
        <div
          class="product-card__text"
        >
          <h2
            class="product-card__name"
          >
            IPhone 14 Pro
          </h2>
          <div
            class="product-card__description"
          >
            Latest iphone, buy it now
          </div>
          <h1
            class="product-card__price"
          />
          <div
            class="product-card__category"
          >
            Электроника
          </div>
        </div>
        <img
          alt="IPhone 14 Pro"
          class="product-card__image"
          data-testid="product-card__image"
          src="/iphone.png"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MainPage render should filter by %category category 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="main-page"
      >
        <h2
          class="main-page__title"
        >
          VK Маркет
        </h2>
        <h3 />
        <div
          class="main-page__parameters"
        >
          <div
            class="categories"
          >
            <div
              class="categories__badge"
            >
              Одежда
            </div>
            <div
              class="categories__badge"
            >
              Для дома
            </div>
            <div
              class="categories__badge categories__badge_selected"
            >
              Электроника
            </div>
          </div>
          <button
            class="sort-button"
            data-testid="sort-button"
          >
            Сортировать 
            по умолчанию
          </button>
        </div>
        <div
          class="product-card"
        >
          <div
            class="product-card__text"
          >
            <h2
              class="product-card__name"
            >
              IPhone 14 Pro
            </h2>
            <div
              class="product-card__description"
            >
              Latest iphone, buy it now
            </div>
            <h1
              class="product-card__price"
            />
            <div
              class="product-card__category"
            >
              Электроника
            </div>
          </div>
          <img
            alt="IPhone 14 Pro"
            class="product-card__image"
            data-testid="product-card__image"
            src="/iphone.png"
          />
        </div>
        <div
          class="product-card"
        >
          <div
            class="product-card__text"
          >
            <h2
              class="product-card__name"
            >
              Принтер
            </h2>
            <div
              class="product-card__description"
            >
              Незаменимая вещь для студента
            </div>
            <h1
              class="product-card__price"
            />
            <div
              class="product-card__category"
            >
              Электроника
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="main-page"
    >
      <h2
        class="main-page__title"
      >
        VK Маркет
      </h2>
      <h3 />
      <div
        class="main-page__parameters"
      >
        <div
          class="categories"
        >
          <div
            class="categories__badge"
          >
            Одежда
          </div>
          <div
            class="categories__badge"
          >
            Для дома
          </div>
          <div
            class="categories__badge categories__badge_selected"
          >
            Электроника
          </div>
        </div>
        <button
          class="sort-button"
          data-testid="sort-button"
        >
          Сортировать 
          по умолчанию
        </button>
      </div>
      <div
        class="product-card"
      >
        <div
          class="product-card__text"
        >
          <h2
            class="product-card__name"
          >
            IPhone 14 Pro
          </h2>
          <div
            class="product-card__description"
          >
            Latest iphone, buy it now
          </div>
          <h1
            class="product-card__price"
          />
          <div
            class="product-card__category"
          >
            Электроника
          </div>
        </div>
        <img
          alt="IPhone 14 Pro"
          class="product-card__image"
          data-testid="product-card__image"
          src="/iphone.png"
        />
      </div>
      <div
        class="product-card"
      >
        <div
          class="product-card__text"
        >
          <h2
            class="product-card__name"
          >
            Принтер
          </h2>
          <div
            class="product-card__description"
          >
            Незаменимая вещь для студента
          </div>
          <h1
            class="product-card__price"
          />
          <div
            class="product-card__category"
          >
            Электроника
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MainPage render should filter by %category category 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="main-page"
      >
        <h2
          class="main-page__title"
        >
          VK Маркет
        </h2>
        <h3 />
        <div
          class="main-page__parameters"
        >
          <div
            class="categories"
          >
            <div
              class="categories__badge"
            >
              Одежда
            </div>
            <div
              class="categories__badge categories__badge_selected"
            >
              Для дома
            </div>
            <div
              class="categories__badge"
            >
              Электроника
            </div>
          </div>
          <button
            class="sort-button"
            data-testid="sort-button"
          >
            Сортировать 
            по умолчанию
          </button>
        </div>
        <div
          class="product-card"
        >
          <div
            class="product-card__text"
          >
            <h2
              class="product-card__name"
            >
              Настольная лампа
            </h2>
            <div
              class="product-card__description"
            >
              Говорят, что ее использовали в pixar
            </div>
            <h1
              class="product-card__price"
            />
            <div
              class="product-card__category"
            >
              Для дома
            </div>
          </div>
          <img
            alt="Настольная лампа"
            class="product-card__image"
            data-testid="product-card__image"
            src="/lamp.png"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="main-page"
    >
      <h2
        class="main-page__title"
      >
        VK Маркет
      </h2>
      <h3 />
      <div
        class="main-page__parameters"
      >
        <div
          class="categories"
        >
          <div
            class="categories__badge"
          >
            Одежда
          </div>
          <div
            class="categories__badge categories__badge_selected"
          >
            Для дома
          </div>
          <div
            class="categories__badge"
          >
            Электроника
          </div>
        </div>
        <button
          class="sort-button"
          data-testid="sort-button"
        >
          Сортировать 
          по умолчанию
        </button>
      </div>
      <div
        class="product-card"
      >
        <div
          class="product-card__text"
        >
          <h2
            class="product-card__name"
          >
            Настольная лампа
          </h2>
          <div
            class="product-card__description"
          >
            Говорят, что ее использовали в pixar
          </div>
          <h1
            class="product-card__price"
          />
          <div
            class="product-card__category"
          >
            Для дома
          </div>
        </div>
        <img
          alt="Настольная лампа"
          class="product-card__image"
          data-testid="product-card__image"
          src="/lamp.png"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MainPage render should filter by %category category 3`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="main-page"
      >
        <h2
          class="main-page__title"
        >
          VK Маркет
        </h2>
        <h3 />
        <div
          class="main-page__parameters"
        >
          <div
            class="categories"
          >
            <div
              class="categories__badge categories__badge_selected"
            >
              Одежда
            </div>
            <div
              class="categories__badge"
            >
              Для дома
            </div>
            <div
              class="categories__badge"
            >
              Электроника
            </div>
          </div>
          <button
            class="sort-button"
            data-testid="sort-button"
          >
            Сортировать 
            по умолчанию
          </button>
        </div>
        <div
          class="product-card"
        >
          <div
            class="product-card__text"
          >
            <h2
              class="product-card__name"
            >
              Костюм гуся
            </h2>
            <div
              class="product-card__description"
            >
              Запускаем гуся, работяги
            </div>
            <h1
              class="product-card__price"
            />
            <div
              class="product-card__category"
            >
              Одежда
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="main-page"
    >
      <h2
        class="main-page__title"
      >
        VK Маркет
      </h2>
      <h3 />
      <div
        class="main-page__parameters"
      >
        <div
          class="categories"
        >
          <div
            class="categories__badge categories__badge_selected"
          >
            Одежда
          </div>
          <div
            class="categories__badge"
          >
            Для дома
          </div>
          <div
            class="categories__badge"
          >
            Электроника
          </div>
        </div>
        <button
          class="sort-button"
          data-testid="sort-button"
        >
          Сортировать 
          по умолчанию
        </button>
      </div>
      <div
        class="product-card"
      >
        <div
          class="product-card__text"
        >
          <h2
            class="product-card__name"
          >
            Костюм гуся
          </h2>
          <div
            class="product-card__description"
          >
            Запускаем гуся, работяги
          </div>
          <h1
            class="product-card__price"
          />
          <div
            class="product-card__category"
          >
            Одежда
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MainPage render should render correctly 1`] = `
<DocumentFragment>
  <div
    class="main-page"
  >
    <h2
      class="main-page__title"
    >
      VK Маркет
    </h2>
    <h3 />
    <div
      class="main-page__parameters"
    >
      <div
        class="categories"
      >
        <div
          class="categories__badge"
        >
          Одежда
        </div>
        <div
          class="categories__badge"
        >
          Для дома
        </div>
        <div
          class="categories__badge"
        >
          Электроника
        </div>
      </div>
      <button
        class="sort-button"
        data-testid="sort-button"
      >
        Сортировать по умолчанию
      </button>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          IPhone 14 Pro
        </h2>
        <div
          class="product-card__description"
        >
          Latest iphone, buy it now
        </div>
        <h1
          class="product-card__price"
        />
        <div
          class="product-card__category"
        >
          Электроника
        </div>
      </div>
      <img
        alt="IPhone 14 Pro"
        class="product-card__image"
        data-testid="product-card__image"
        src="/iphone.png"
      />
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Костюм гуся
        </h2>
        <div
          class="product-card__description"
        >
          Запускаем гуся, работяги
        </div>
        <h1
          class="product-card__price"
        />
        <div
          class="product-card__category"
        >
          Одежда
        </div>
      </div>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Настольная лампа
        </h2>
        <div
          class="product-card__description"
        >
          Говорят, что ее использовали в pixar
        </div>
        <h1
          class="product-card__price"
        />
        <div
          class="product-card__category"
        >
          Для дома
        </div>
      </div>
      <img
        alt="Настольная лампа"
        class="product-card__image"
        data-testid="product-card__image"
        src="/lamp.png"
      />
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Принтер
        </h2>
        <div
          class="product-card__description"
        >
          Незаменимая вещь для студента
        </div>
        <h1
          class="product-card__price"
        />
        <div
          class="product-card__category"
        >
          Электроника
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
